lib/SPIRVProducerPass.cpp

```diff
@@ -6376,3 +6481,107 @@ uint64_t SPIRVProducerPass::GetTypeAllocSize(Type *type, const DataLayout &DL) {
+  return v;
+}
+
+bool SPIRVProducerPass::sameResource(Value *lhs, Value *rhs) const {
+  if (auto *lhs_call = dyn_cast<CallInst>(lhs)) {
+    if (auto *rhs_call = dyn_cast<CallInst>(rhs)) {
+      if (lhs_call->getCalledFunction()->getName().startswith(
+              clspv::ResourceAccessorFunction()) &&
+          rhs_call->getCalledFunction()->getName().startswith(
+              clspv::ResourceAccessorFunction())) {
+        // For resource accessors, match descriptor set and binding.
+        if (lhs_call->getOperand(0) == rhs_call->getOperand(0) &&
+            lhs_call->getOperand(1) == rhs_call->getOperand(1))
+          return true;
+      } else if (lhs_call->getCalledFunction()->getName().startswith(
+                     clspv::WorkgroupAccessorFunction()) &&
+                 rhs_call->getCalledFunction()->getName().startswith(
+                     clspv::WorkgroupAccessorFunction())) {
+        // For workgroup resources, match spec id.
+        if (lhs_call->getOperand(0) == rhs_call->getOperand(0))
+          return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+bool SPIRVProducerPass::selectFromSameObject(Instruction *inst) {
+  assert(inst->getType()->isPointerTy());
+  assert(GetStorageClass(inst->getType()->getPointerAddressSpace()) ==
+         spv::StorageClassStorageBuffer);
+  const bool hack_undef = clspv::Option::HackUndef();
+  if (auto *select = dyn_cast<SelectInst>(inst)) {
+    auto *true_base = GetBasePointer(select->getTrueValue());
+    auto *false_base = GetBasePointer(select->getFalseValue());
+
+    if (true_base == false_base)
+      return true;
+
+    // If either the true or false operand is a null, then we satisfy the same
+    // object constraint.
+    if (auto *true_cst = dyn_cast<Constant>(true_base)) {
+      if (true_cst->isNullValue() || (hack_undef && isa<UndefValue>(true_base)))
+        return true;
+    }
+
+    if (auto *false_cst = dyn_cast<Constant>(false_base)) {
+      if (false_cst->isNullValue() ||
+          (hack_undef && isa<UndefValue>(false_base)))
+        return true;
+    }
+
+    if (sameResource(true_base, false_base))
+      return true;
+  } else if (auto *phi = dyn_cast<PHINode>(inst)) {
+    Value *value = nullptr;
+    bool ok = true;
+    for (unsigned i = 0; ok && i != phi->getNumIncomingValues(); ++i) {
+      auto *base = GetBasePointer(phi->getIncomingValue(i));
+      // Null values satisfy the constraint of selecting of selecting from the
+      // same object.
+      if (!value) {
+        if (auto *cst = dyn_cast<Constant>(base)) {
+          if (!cst->isNullValue() && !(hack_undef && isa<UndefValue>(base)))
+            value = base;
+        } else {
+          value = base;
+        }
+      } else if (base != value) {
+        if (auto *base_cst = dyn_cast<Constant>(base)) {
+          if (base_cst->isNullValue() || (hack_undef && isa<UndefValue>(base)))
+            continue;
+        }
+
+        if (sameResource(value, base))
```

Comment:
Slightly confusing.  I would have written.
`ok = ok && sameResource(value,base)` and avoided the continue.